### PR TITLE
BUG: support psycopg3 in .to_postgis()

### LINKED ISSUE #3061: BUG: `to_postgis()` not compatible with psycopg v3
- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of geopandas.

- [x] (optional) I have confirmed this bug exists on the main branch of geopandas.

---

#### Problem description

The `to_postgis()` docs state:

> This method requires SQLAlchemy and GeoAlchemy2, and a PostgreSQL Python driver (e.g. psycopg2) to be installed.

I understand the "e.g." to mean that other drivers like psycopg (version 3) are/should be possible. 

However, `to_postgis()` or rather `_psql_insert_copy()` uses psycopg2's `copy_expert()` method:
https://github.com/geopandas/geopandas/blob/8f2b4cc1b2cbb62aed716c9a187e800e79df3cd4/geopandas/io/sql.py#L328

`copy_expert()` has been deprecated in psycopg in favor of [a single `copy()` method](https://www.psycopg.org/psycopg3/docs/basic/from_pg2.html#copy-is-no-longer-file-based). Hence, `to_postgis()` fails when used with a psycopg-enabled sqlAlchemy engine.

(More generally the use of a psycopg2 method means that `to_postgis()` only supports _that_ driver and not with any of the [other drivers supported by sqlAlchemy](https://docs.sqlalchemy.org/en/20/dialects/postgresql.html#dialect-postgresql), though I cannot say anything about their practical relevance. Except for [psycopg 3](https://www.psycopg.org/psycopg3/), the successor of psycopg2).

### PR DESCRIPTION
**Pull request by MrAlecJohnson**
Closes #3061, which found that `.to_postgis()` uses `copy_expert`, which is a method specific to psycopg2. Some users may want to use the more recent psycopg (also known as psycopg3) package instead. A reply to the issue report asked for a pull request adding psycopg support, so I've put this one together as I've been encountering the same problem. 

## Changes

This pull request makes `.to_postgis()` use either `.copy()` (psycopg) or `copy_expert()` depending on which is available on the cursor object. It does this by adding an attribute check to `._psql_insert_copy()`, which is passed to `.to_sql` as the 'method' argument.

The biggest changes are to the tests for `io/sql`, where it parameterises all tests to run with both psycopg and psycopg2 to ensure compatibility for both. 

It also adds psycopg to the docs and test builds. 

## Questions

I'm not sure if you'd prefer to keep one out of psycopg2/psycopg as the 'main' protocol. I've treated them pretty much evenly but put psycopg first in the `to_postgis()` check as it's the more recent package.

### CODE REVIEW

**Comment by MrAlecJohnson** [comment]
(Hopefully this is relatively close to ready, but I've closed it till I can get back to it and get all the tests working, so it doesn't clutter your pull requests in the meantime)

**Comment by MrAlecJohnson** [comment]
Problems with test runners seem largely resolved, but there's still a couple I haven't got my head round. Will leave in draft till I can get back to this and work out what's going wrong

**Comment by m-richards** [comment]
@MrAlecJohnson I haven't looked through your PR in detail, but it might be handy to know
* the 3.11 dev enviroment builds against dev versions of packages, and if you see failures that look unrelated to what you're working on, they probably are (and we can clarify if they are and you are unsure)
* The tests.yaml handles postgis separately (you may have already seen this), so the postgis tests actually only run in `310-pd20-conda-forge.yaml` and `311-latest-conda-forge.yaml`.
Thanks for the draft contribution, I'll have a look in some detail hopefully soon.

**Comment by MrAlecJohnson** [comment]
Ah, okay, I see the issue - it's about dependency versions, so @m-richards it would be great to get your (or anyone else's!) views on how I should approach this. I don't want to mess with the builds unnecessarily.  

1. there's a tiny error in the Pandas compatibility xfails on [test_duplicate_geometry_column_fails](https://github.com/geopandas/geopandas/blob/e996129c5f77fb7aae1d7bbf76dee3c7678a8ecb/geopandas/io/tests/test_sql.py#L743) - this will xfail on 2.0.0 and 2.0.1, but xpass on 2.0.2 and 2.0.3 - and `310-pd20-conda-forge` pins to 2.0 without specifying minor version. I think the easiest fix here is to pin a specific minor version in this build - does that sound reasonable? Or would you rather I make the pandas compatbility checks more specific?

2. I failed to account for psycopg3 requiring sqlalchemy>=2.0. Rather than changing any of the sqlalchemy versions in the builds, I suspect I should just only include psycopg3 when there's already a sqlalchemy above 2.0. Does that also sound sensible? The new psycopg3 postgis tests run nicely on 311-latest-conda-forge, where sqlalchemy>=2.0

**Comment by m-richards** [comment]
> 1. there's a tiny error in the Pandas compatibility xfails on [test_duplicate_geometry_column_fails](https://github.com/geopandas/geopandas/blob/e996129c5f77fb7aae1d7bbf76dee3c7678a8ecb/geopandas/io/tests/test_sql.py#L743) - this will xfail on 2.0.0 and 2.0.1, but xpass on 2.0.2 and 2.0.3 - and `310-pd20-conda-forge` pins to 2.0 without specifying minor version. I think the easiest fix here is to pin a specific minor version in this build - does that sound reasonable? Or would you rather I make the pandas compatbility checks more specific?

I would prefer to fix in in the code than in the build deps - you'll notice this is only failing on this branch and not on main, and I presume that's because adding psycopg3 to the yml is causing conda to resolve to a different version of pandas. I think you could define a compat.PANDAS_GE_202 analogous to the others and update the guard in the test (https://github.com/pandas-dev/pandas/pull/53118 is where this was fixed on the pandas side and it lines up with the versions you have checked xfail/ xpass)

> 2. I failed to account for psycopg3 requiring sqlalchemy>=2.0. Rather than changing any of the sqlalchemy versions in the builds, I suspect I should just only include psycopg3 when there's already a sqlalchemy above 2.0. Does that also sound sensible? The new psycopg3 postgis tests run nicely on 311-latest-conda-forge, where sqlalchemy>=2.0

I agree this makes sense - at least for this PR. We only have two dev builds running the postgres tests, so it's good to keep one on sqlalchemy <2 and one on 2+. (Although I think this point 1. of yours will be resolved without you needing to make any code changes, so I supose that's fine too :))

We are at the point where we can probably consider dropping sqlalchemy 1.4 support - but I think thats a separate topic, because pandas didn't support sql 2 until pandas 2.

**Comment by MrAlecJohnson** [comment]
Those approaches sound great, thank you. 

Belatedly realised my 'skip tests for packages not installed' approach won't work with the 'TESTS SKIPPED, FAILING' in the test workflow, so I'll have to come back to that, hopefully soon!

**Comment by MrAlecJohnson** [comment]
I think this is now ready for review. There are still 2 runners failing, but as far as I can tell the failures are unconnected to these changes. Apologies if I've got that wrong! 

For the `test_sql` tests I've set them to only run on versions of psycopg that are installed, so if only one package is installed only one set of tests is run. If neither psycopg nor psycopg2 is installed, no sql-based tests run, but there are still skips, so they will still set off the 'Fail if tests are skipped' condition in the workflow.

File: geopandas/io/tests/test_sql.py
Comment (m-richards) [inline geopandas/io/tests/test_sql.py:58]:
```suggestion
def engine_psycopg(psycopg_package_name: str):
```
I think the fact that this works for multiple versions is implied by the fact that it takes a package name as an argument.

File: geopandas/io/tests/test_sql.py
Comment (m-richards) [inline geopandas/io/tests/test_sql.py:42]:
```suggestion
def connection_psycopg(psycopg_package_name: str):
```

File: geopandas/io/tests/test_sql.py
Comment (m-richards) [inline geopandas/io/tests/test_sql.py:153]: It'd be nice if we didn't need these engine/connection 2/3 variants but I don't think that's easy when we're hooking into the fixture setup time.

**Review by m-richards — COMMENTED** [review]
Thanks @MrAlecJohnson for this. The actual code changes look fine, and I've been a little slow in reviewing since there's a fair bit of complexity going on in the tests. I know that parametrised fixtures can be quite elegant and efficient but it does make this code much harder to understand at a glance (which is not ideal from a maintenance perspective).

I'm not a pytest expert so I couldn't see obviously how to simplify this without looking at the code directly. I've pushed a commit with some changes here: https://github.com/geopandas/geopandas/pull/3205/commits/fafc8fa27046d9991b6847395ab86808b34f5299 let me know what you think. I think it's a little clearer to not use the higher order `parametrised_fixture` everywhere. The diff with main is also slightly cleaner.

File: geopandas/io/tests/test_sql.py
Comment (MrAlecJohnson) [reply geopandas/io/tests/test_sql.py:153]: Hmm, yes, they are pretty unclear and would be great to remove them. But this was the neatest I could come up with without creating a new engine/connection each time. I will have another look and see if there's a better way...

**Comment by MrAlecJohnson** [comment]
@m-richards Thank you for those suggestions. All incorporated now, along with a much cleaner test setup and much smaller diff. I've managed to get rid of those unnecessary extra fixtures now. I hope this looks more useful!

File: CHANGELOG.md
Comment (m-richards) [inline CHANGELOG.md:63]: Don't overly mind, but I'd this could be grouped as a new feature, as it's new functionality (I know the original issue is tagged as a bug though)

File: doc/source/getting_started/install.rst
Comment (m-richards) [inline doc/source/getting_started/install.rst:156]:
```suggestion
- `psycopg2`_ (optional; for PostGIS connection - older version of `psycopg` library)
```

**Review by m-richards — APPROVED** [review]
Thanks @MrAlecJohnson this is in a pretty good spot now (quite happy with the fixture cleanup from your last changes, it's a lot neater).

Will need a second pair of eyes to review before it is merged but I'm happy from my perspective.

File: CHANGELOG.md
Comment (martinfleis) [reply CHANGELOG.md:63]: It is more a compatibility note imho.

**Review by martinfleis — COMMENTED** [review]
Shall we maybe switch from psycopg2 to psycopg in optional-dependencies in pyproject.toml? this psycopg->psycopg2->psycopg versioning is so confusing I am not sure what is the proper dependency.

**Comment by m-richards** [comment]
> Shall we maybe switch from psycopg2 to psycopg in optional-dependencies in pyproject.toml? this psycopg->psycopg2->psycopg versioning is so confusing I am not sure what is the proper dependency.

Yeah I suppose it should be `psycopg` now.

File: CHANGELOG.md
Comment (MrAlecJohnson) [reply CHANGELOG.md]: Thanks! I've moved it to compatibility in a0e90cc

File: CHANGELOG.md
Comment (martinfleis) [inline CHANGELOG.md:81]:
```suggestion
- Fixes for compatibility with psycopg (#3167).
```

**Review by martinfleis — APPROVED** [review]

**Comment by martinfleis** [comment]
Thanks @MrAlecJohnson!

**Comment by rs-kersten** [comment]
How comes that this is still not in the pip package (0.14.4) although merged in main?

**Comment by martinfleis** [comment]
@rs-kersten  main is targeted for 1.0, 0.14.x lives in a branch and this was not back ported when doing 0.14.4 release.

**Comment by rs-kersten** [comment]
Oh okay, thanks for the clarification

**Pull request merged**
